README.md
UWSGI COOKBOOK
==============
[![Build Status](https://secure.travis-ci.org/arktos65/uwsgi-cookbook.png)](http://travis-ci.org/arktos65/uwsgi-cookbook)

uWSGI is a high performance application server framework for Python, Ruby, and other applications.  This cookbook
compiles from source a modular installation of uWSGI.  Using this approach allows for a tailored installation using
wrapper cookbooks.

The Debian configure recipe is somewhat reverse engineered from the uwsgi apt package, but set up in such a way to
allow flexible configuration.

## Supported Platforms

Supported platforms:  Ubuntu Server 14.04, 16.04

Resources
---------
The resources that ship with this cookbook allow you to manage uWSGI applications from other cookbooks, particularly
wrapper cookbooks.

## Example

    uwsgi_application 'some_application.conf' do
        source 'application_template.conf'
        action [:create, :enable]
    end
    
The application name should not include a path, only the target file name.  The attribute `node['uwsgi']['uwsgi']['config']['directories']['apps_available']`
contains the path where the target will be copied to.  When using the :enable action, the attribute
`node['uwsgi']['config']['directories']['apps_enabled']` specifies where the symlink will be created.  

In a wrapper cookbook, if you wish to change the default locations of the apps_available and apps_enabled attributes,
use the node.set in your recipe before using the uwsgi_application resource. Your application configuration files
should be stored in the templates directory.

## Attributes

    default['uwsgi']['version'] - Version to download and install.
    default['uwsgi']['download_url'] - URL of source code.
    default['uwsgi']['service'] - Service name to be created.
    
    default['uwsgi']['core']['binary'] - File name of uWSGI core binary.
    default['uwsgi']['core']['directory'] - Path where core binary and plugins will be installed.
    
    default['uwsgi']['config']['directories']['etc'] = "/etc/uwsgi"
    default['uwsgi']['config']['directories']['apps_available'] = "/etc/uwsgi/apps-available"
    default['uwsgi']['config']['directories']['apps_enabled'] = "/etc/uwsgi/apps-enabled"
    default['uwsgi']['config']['directories']['logs'] = "/var/log/uwsgi"
    default['uwsgi']['config']['directories']['logs_app'] = "/var/log/uwsgi/app"
    default['uwsgi']['config']['directories']['run'] = "/run/uwsgi"
    default['uwsgi']['config']['directories']['run_app'] = "/var/run/uwsgi/app"
    default['uwsgi']['config']['directories']['share'] = "/usr/share/uwsgi"
    default['uwsgi']['config']['directories']['share_init'] = "/usr/share/uwsgi/init"
    default['uwsgi']['config']['directories']['share_conf'] = "/usr/share/uwsgi/conf"
    
Control which plugins are compiled by setting each plugins' :compile attribute to true or false.  See the 
`attributes/build-plugins.rb` for a list of available plugins.

## Usage

### uwsgi::default

Include `uwsgi` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[uwsgi::default]"
  ]
}
```

License & Authors
-----------------
- Author:: Sean M. Sullivan (<sean@barriquesoft.com>)

```text
Copyright:: 2014-2017 Pulselocker, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```



CHANGELOG.md
UWSGI COOKBOOK CHANGELOG
========================

1.3.5 (2017-11-30)
------------------
- Updated contact information in README file.

1.3.4 (2017-09-28)
------------------
- Added Ubuntu 16.04 to Travis CI testing.
- Changed Travis testing transport from Docker to Dokken.
- Added `linter` task to `Rakefile`.

1.3.3 (2017-08-24)
------------------
- Refactored code to prevent resource duplication.
- Updated copyright information.

1.3.2 (2017-08-15)
------------------
- Restructured directory to work with Berkshelf.

1.3.1 (2017-08-10)
------------------
- Fixed license metadata to satisfy Foodcritic.
- [RESOLVED] Python command missing on Ubuntu 16.04 test #16
- Default core build is now uWSGI 2.0.15.
- Updated Travis test harness.
- Added Rubocop style linter support.
- Added Test Kitchen testing in Travis CI using Docker.
- Converted hashes to Ruby 1.9 syntax throughout source code.

1.3.0 (2017-08-09)
------------------
- Deprecated `python` cookbook in favor of `poise-python`.
- Refactor `metadata.rb` for correct cookbook dependencies.
- Pinned cookbook to support Chef 12.
- Added testing framework.

1.2.1 (2017-02-08)
------------------
- Fixed issue #13 where service fails to start with `$ service uwsgi start` command.
- Modified many file permissions from 0755 to 0644 for improved security.

1.2.0 (2017-01-10)
------------------
- Added Ubuntu 16.04 to Test Kitchen.
- Bumped `rsyslog` cookbook version.
- Bumped `build-essential` cookbook version.
- Bumped `uwsgi` source version from 2.0.9 to 2.0.14.

1.1.3 (2016-03-15)
------------------
- Updated Travis configuration to support Ruby 2.2.1.
- Modified Rakefile to require Ruby 2.2.1 or higher for foodcritic.
- Changed package directive in default recipe per `foodcritic` recommendation.
- Updated `metadata.rb` to comply with Chef 11/12 compatibility per `foodcritic`.

1.1.2 (2015-11-16)
------------------
- Correct attributes & allow logrotate to process file.

1.1.1 (2015-06-25)
------------------
- Installs libpcre3 and libpcre3-dev packages on Debian flavored distributions.
- Changed build_uwsgi_moduler.ini file to support PCRE and internal routing by default for modular build.

1.1.0
-----
- Includes support for Emperor server configuration.  Set `node['uwsgi']['emperor']['enable']` to true.
- Deprecated Test Kitchen suite for Ubuntu 12.04.

1.0.3
-----
- Improvements to the packaging of uWSGI plugins.

1.0.2
-----
- Enable `rsyslog` plugin support in default configuration.
- Default uWSGI core build is now `v2.0.9`.

1.0.1
-----
- Added Travis CI testing support.

1.0.0
-----
- Version bump for production release.

0.2.0
-----
- Added LWRP for uWSGI application management.

0.1.1
-----
- Package dependencies added.

0.1.0
-----
- Initial development release of uWSGI cookbook.
- Builds core binary and installs in runtime directory (default /usr/local/bin/uwsgi).



recipes/prepare.rb
# frozen_string_literal: true

#
# Cookbook Name:: uwsgi
# Recipe:: _prepare
#
# Copyright 2014-2017, Pulselocker, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

###
# Prepare the environment for installation
###

if node['platform_family'] == 'debian'
  node.override['apt']['compile_time_update'] = true
  include_recipe 'apt'
end
include_recipe 'rsyslog'
include_recipe 'build-essential'
include_recipe 'poise-python'

# Workaround for bug in Python 2.7 apt package (Issue #16)
unless File.exist?('/usr/bin/python')
  link '/usr/bin/python' do
    to '/usr/bin/python2.7'
    owner 'root'
    group 'root'
  end
  link '/usr/bin/python2' do
    to '/usr/bin/python2.7'
    owner 'root'
    group 'root'
  end
end

# Add other dependencies
if node['platform_family'] == 'debian' && node['uwsgi']['pcre']['enable']
  package 'libpcre3' do
    action :install
  end
  package 'libpcre3-dev' do
    action :install
  end
end

# Download the source code
remote_file "#{Chef::Config[:file_cache_path]}/uwsgi-#{node['uwsgi']['version']}.tar.gz" do
  source "#{node['uwsgi']['download_url']}/uwsgi-#{node['uwsgi']['version']}.tar.gz"
  action :create_if_missing
end

bash "extract_uwsgi_#{node['uwsgi']['version']}_source" do
  cwd Chef::Config[:file_cache_path]
  code <<-EOH
    tar -zxvf uwsgi-#{node['uwsgi']['version']}.tar.gz
  EOH
end

directory node['uwsgi']['buildconf'] do
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end

# Create directories for uWSGI binaries
directory node['uwsgi']['core']['directory'] do
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end
directory node['uwsgi']['plugins']['root'] do
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end
directory node['uwsgi']['plugins']['directory'] do
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end
